indexing-service/src/main/java/io/druid/indexing/common/actions/SegmentAllocateAction.java

```diff
@@ -0,0 +1,268 @@
+        preferredSegmentGranularity,
+        "preferredSegmentGranularity"
+    );
+    this.sequenceName = Preconditions.checkNotNull(sequenceName, "sequenceName");
+    this.previousSegmentId = previousSegmentId;
+  }
+
+  @JsonProperty
+  public String getDataSource()
+  {
+    return dataSource;
+  }
+
+  @JsonProperty
+  public DateTime getTimestamp()
+  {
+    return timestamp;
+  }
+
+  @JsonProperty
+  public QueryGranularity getQueryGranularity()
+  {
+    return queryGranularity;
+  }
+
+  @JsonProperty
+  public Granularity getPreferredSegmentGranularity()
+  {
+    return preferredSegmentGranularity;
+  }
+
+  @JsonProperty
+  public String getSequenceName()
+  {
+    return sequenceName;
+  }
+
+  @JsonProperty
+  public String getPreviousSegmentId()
+  {
+    return previousSegmentId;
+  }
+
+  @Override
+  public TypeReference<Optional<SegmentIdentifier>> getReturnTypeReference()
+  {
+    return new TypeReference<Optional<SegmentIdentifier>>()
+    {
+    };
+  }
+
+  @Override
+  public Optional<SegmentIdentifier> perform(
+      final Task task,
+      final TaskActionToolbox toolbox
+  ) throws IOException
+  {
+    if (!task.getDataSource().equals(dataSource)) {
+      throw new IAE("Task dataSource must match action dataSource, [%s] != [%s].", task.getDataSource(), dataSource);
+    }
+
+    final IndexerMetadataStorageCoordinator msc = toolbox.getIndexerMetadataStorageCoordinator();
+
+    // 1) if something overlaps our timestamp, use that
+    // 2) otherwise try preferredSegmentGranularity & going progressively smaller
+
+    final List<Interval> tryIntervals = Lists.newArrayList();
+
+    final Interval rowInterval = new Interval(
+        queryGranularity.truncate(timestamp.getMillis()),
+        queryGranularity.next(queryGranularity.truncate(timestamp.getMillis()))
+    );
+
+    final Set<DataSegment> usedSegmentsForRow = ImmutableSet.copyOf(
+        msc.getUsedSegmentsForInterval(dataSource, rowInterval)
+    );
+
+    if (usedSegmentsForRow.isEmpty()) {
+      // No existing segments for this row, but there might still be nearby ones that conflict with our preferred
+      // segment granularity. Try that first, and then progressively smaller ones if it fails.
+      for (Granularity gran : granularitiesFinerThan(preferredSegmentGranularity)) {
+        tryIntervals.add(gran.bucket(timestamp));
+      }
+    } else {
+      // Existing segment(s) exist for this row; use the interval of the first one.
+      tryIntervals.add(usedSegmentsForRow.iterator().next().getInterval());
+    }
+
+    for (final Interval tryInterval : tryIntervals) {
+      if (tryInterval.contains(rowInterval)) {
+        log.debug(
+            "Trying to allocate pending segment for rowInterval[%s], segmentInterval[%s].",
+            rowInterval,
+            tryInterval
+        );
+        final TaskLock tryLock = toolbox.getTaskLockbox().tryLock(task, tryInterval).orNull();
```

Comment:
note: this assumes, tasks in different replication set "share" the lock so that they both will be able to get lock on same interval if they received events with same or close enough timestamp.